### PR TITLE
[tune] Make PBT Quantile fraction configurable

### LINKED ISSUE #4881: [tune] Make PBT Quantile fraction configurable
### Describe the problem

Currently PBT has the quantile fraction hardcoded to 25% (that is trials from the top 25% get transferred to the bottom 25%). 
It would be nice if this is configurable.

I have been bothering you guys with comments/issues for a while, so I'd be happy to use this as my first issue to contribute, if there are no objections from you guys or any reasons you see that this should not be configurable. :)

### PR DESCRIPTION
**Pull request by timonbimon**
## What do these changes do?
Make the PBT quantile fraction configurable.

## Related issue number

Closes #4881 

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

## A few comments as a first-time contributor

Please take these with a grain of salt. There are usually good historical reasons for the way things are setup, but I thought it might be good to still voice my opinion so that you guys get a fresh view on things. :) 

### Tests
Running tests seems to fail out of the box:
- Need to install ‘gym’, ‘pandas’, ‘opencv-python’ (cv2)

To get tests to pass you also need to
- Install ‘requests’, ’tensorflow’, ’tabulate'

This should at least be mentioned in the docs I think (or better have all dev requirements listed somewhere to make them easily installable, e.g. [poetry](https://poetry.eustace.io/) has a dev section in the pyproject.toml or if one really still wants to go with default pip one could have a requirements-dev.txt somewhere).

- all the tune tests seem to be written with unittest => going forward it would probably be nice to transition to pytest (I think this was already done in other parts of the ray repo)

### Linting, auto-formatting etc.
- Imho it would be nicer to have these run via pre-commit hooks instead of manually having to run `ray/scripts/format.sh`. The more automation the better. :) (Minor comment: it seems there are two identical re-formatting scripts checked-in to the code base - one at ci/travis/format.sh and the other at scripts/format.sh - this seems like a bad idea and violation of DRY)
- I also think black is the better auto-formatter for Python (https://news.ycombinator.com/item?id=17155048), but I understand how switching the auto-formatter is probably not a priority ;-)

### CODE REVIEW

**Comment by timonbimon** [comment]
@richardliaw

File: python/ray/tune/schedulers/pbt.py [hunk 1]
```diff
@@ -228,6 +234,7 @@ def _log_config_on_step(self, trial_state, new_state, trial,
 
         For each step, logs: [target trial tag, clone trial tag, target trial
         iteration, clone trial iteration, old config, new config].
+
```
Comment (richardliaw) [inline python/ray/tune/schedulers/pbt.py]:
```suggestion
```

Comment (timonbimon) [reply python/ray/tune/schedulers/pbt.py]: hmm it says your comment was marked as resolved. do you want me to remove the empty lines at the end of docstrings again?

Comment (richardliaw) [reply python/ray/tune/schedulers/pbt.py]: Ah I checked the style guide and what you have is right.

**Comment by timonbimon** [comment]
hmm the CI error seems to say

```
~/build/ray-project/ray/thirdparty/pkg/redis ~/build/ray-project/ray/python

+make

make: *** No targets specified and no makefile found.  Stop.
```

Apparently there is a Makefile missing...
Not sure what to do with this.

**Comment by richardliaw** [comment]
I think the dead test is fine. Would you mind adding a test in `tests/test_trial_scheduler`?

Thanks for your feedback! Replies below:

RE: missing dependencies, yeah that's a good point... Would be good to have a `requirements-dev.txt`. I think one first pass would be just to add it under `tune/` for a Tune-specific requirements list (and also mention it on the contributor guide).

RE: pytest, I have no strong opinion, and also RLlib isn't planning to transition to pytest. cc @ericl 

RE: black; I've seen black but I think the cost of doing switching autoformatters (it covers a bunch of git history and also not particularly necessary) at this point outweighs the pros :) 

RE: duplication of `format.sh` - I think we could remove `scripts/format.sh` and change any references of it to the one in `ci/`

**Comment by ericl** [comment]
In fact, scripts/ and ci/travis are connected by symbolic link, so there is no duplication going on.

**Comment by ericl** [comment]
Re pytest, I would rather we remove the unnecessary dependency on pytest!

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
Re: pytest, I've been bitten too much by it to want to use it more. IMO we should consider breaking down tune into more, smaller test files, and getting rid of any pytest runners in favor of simply having bazel manage the execution.

**Comment by timonbimon** [comment]
Would be curious to hear how pytest has bitten you in the past? :)
Also just made a minor change to make sure the quantiles do not overlap (which was the case before for a list of trials with uneven length and quantile_fraction 0.5)

**Pull request merged**